x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/planner/EsqlExpressionTranslators.java

```diff
@@ -531,21 +532,31 @@ private static RangeQuery translate(Range r, TranslatorHandler handler) {
     public static class MatchFunctionTranslator extends ExpressionTranslator<Match> {
         @Override
         protected Query asQuery(Match match, TranslatorHandler handler) {
-            return new MatchQuery(match.source(), ((FieldAttribute) match.field()).name(), match.queryAsText());
+            Expression field = match.field();
+            if (field instanceof FieldAttribute fieldAttribute) {
+                String fieldName = fieldAttribute.name();
+                if (fieldAttribute.field() instanceof MultiTypeEsField multiTypeEsField) {
+                    // If we have multiple field types, we allow the query to be done, but getting the underlying field name
+                    fieldName = multiTypeEsField.getName();
+                }
+                return new MatchQuery(match.source(), fieldName, match.queryAsObject());
+            }
+
+            throw new IllegalArgumentException("Match must have a field attribute as the first argument");
```

Comment:
Thank you @carlosdelest ! I'm on the fence about whether keeping these(defecting the real field name and throw `IllegalArgumentException`) on the data node or on the coordinator node(inside the `Match` function itself), so that we can keep `MatchFunctionTranslator` clean here, `throw new IllegalArgumentException` on data node caught my eyes. Is this something can be done safely on the coordinator node?

I have tried this, and the tests related to multi-typed fields that I can think of return as expected. I'd suggest to have more tests to cover more data types besides `int` and `long` with multi-type fields as part of the regression tests - `double`, `unsigned`, `boolean`, strings, without having them in the regression tests it is hard to tell when things change. `semantic_text` will be supported by `match` soon, depending on which PR merges first, we should have tests to cover it as well.